README
NAME
    Class::Null - Implements the Null Class design pattern

VERSION
    version 2.101420

SYNOPSIS
      use Class::Null;

      # some class constructor and accessor declaration here

      sub init {
        my $self = shift;
        # ...
        $self->log(Class::Null->new);
        # ...
      }

      sub do_it {
        my $self = shift;
        $self->log->log(level => 'debug', message => 'starting to do it');
        # ...
        $self->log->log(level => 'debug', message => 'still doing it');
        # ...
        $self->log->log(level => 'debug', message => 'finished doing it');
      }

DESCRIPTION
    This class implements the Null Class design pattern.

    Suppose that methods in your object want to write log messages to a log
    object. The log object is possibly stored in a slot in your object and
    can be accessed using an accessor method:

      package MyObject;

      use base 'Class::Accessor';
      __PACKAGE__->mk_accessors(qw(log));

      sub do_it {
        my $self = shift;
        $self->log->log(level => 'debug', message => 'starting to do it');
        ...
        $self->log->log(level => 'debug', message => 'still doing it');
        ...
        $self->log->log(level => 'debug', message => 'finished doing it');
      }

    The log object simply needs to have a "log()" method that accepts two
    named parameters. Any class defining such a method will do, and
    "Log::Dispatch" fulfills that requirement while providing a lot of
    flexibility and reusability in handling the logged messages.

    You might want to log messages to a file:

      use Log::Dispatch;

      my $dispatcher = Log::Dispatch->new;

      $dispatcher->add(Log::Dispatch::File->new(
        name      => 'file1',
        min_level => 'debug',
        filename  => 'logfile'));

      my $obj = MyObject->new(log => $dispatcher);
      $obj->do_it;

    But what happens if we don't define a log object? Your object's methods
    would have to check whether a log object is defined before calling the
    "log()" method. This leads to lots of unwieldy code like

      sub do_it {
        my $self = shift;
        if (defined (my $log = $self->log)) {
          $log->log(level => 'debug', message => 'starting to do it');
        }
        ...
        if (defined (my $log = $self->log)) {
          $log->log(level => 'debug', message => 'still doing it');
        }
        ...
        if (defined (my $log = $self->log)) {
          $log->log(level => 'debug', message => 'finished doing it');
        }
      }

    The proliferation of if-statements really distracts from the actual call
    to "log()" and also distracts from the rest of the method code. There is
    a better way. We can ensure that there is always a log object that we
    can call "log()" on, even if it doesn't do very much (or in fact,
    anything at all).

    This object with null functionality is what is called a null object. We
    can create the object the usual way, using the "new()" constructor, and
    call any method on it, and all methods will do the same - nothing.
    (Actually, it always returns the same "Class::Null" singleton object,
    enabling method chaining.) It's effectively a catch-all object. We can
    use this class with our own object like this:

      package MyObject;

      use Class::Null;

      # some class constructor and accessor declaration here

      sub init {
        my $self = shift;
        ...
        $self->log(Class::Null->new);
        ...
      }

      sub do_it {
        my $self = shift;
        $self->log->log(level => 'debug', message => 'starting to do it');
        ...
        $self->log->log(level => 'debug', message => 'still doing it');
        ...
        $self->log->log(level => 'debug', message => 'finished doing it');
      }

    This is only one example of using a null class, but it can be used
    whenever you want to make an optional helper object into a mandatory
    helper object, thereby avoiding unnecessarily complicated checks and
    preserving the transparency of how your objects are related to each
    other and how they call each other.

    Although "Class::Null" is exceedingly simple it has been made into a
    distribution and put on CPAN to avoid further clutter and repetitive
    definitions.

METHODS
  new
    Returns the singleton null object.

  Any other method
    Returns another singleton null object so method chaining works.

OVERLOADS
    Boolean context
        In boolean context, a null object always evaluates to false.

    Numeric context
        When used as a number, a null object always evaluates to 0.

    String context
        When stringified, a null object always evaluates to the empty
        string.

INSTALLATION
    See perlmodinstall for information and options on installing Perl
    modules.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <http://rt.cpan.org/Public/Dist/Display.html?Name=Class-Null>.

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see <http://search.cpan.org/dist/Class-Null/>.

    The development version lives at
    <http://github.com/hanekomu/Class-Null/>. Instead of sending patches,
    please fork this project using the standard git and github
    infrastructure.

AUTHOR
      Marcel Gruenauer <marcel@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2005 by Marcel Gruenauer.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension Class-Null

2.101420  2010-05-22 23:09:16 Europe/Vienna
    - 'use parent' instead of 'use base'

2.100830  2010-03-24 11:44:23 Europe/Vienna
    - converted the distribution to Dist::Zilla-style

1.11  Mon 2009.12.07 22:18:14 CET (Marcel Gruenauer <marcel@cpan.org>)
     - reduced dependencies using author tests
     - better dist engineering (INSTALL, README, README.mkdn etc.)

1.10  Fri Nov  7 11:43:47 CET 2008 (Marcel Gruenauer <marcel@cpan.org>)
     - Changed dist style and Changes back to standard. People didn't like it -
       the nail that sticks out gets hammered down.
     - Added standard test files; this will also help with CPANTS.
     - Fixed a bug where Class::Null->new->new didn't work (thanks rabbit,
       rt.cpan.org bug #34244)

1.09  Thu, 29 May 2008 12:01:38 -0000 (Marcel Gruenauer <marcel@cpan.org>)
     - Converted Changes file to YAML style
     - .shipit: added Twitter step
     - Makefile.PL: added auto_install() and process_templates()
     - lib/*: converted to template
     - updated MANIFEST
     - added t/perlcriticrc
     - updated MANIFEST
     - .shipit: fixed svk.tagpattern
     - tags: NEWFEATURE

1.08  Sat, 20 Oct 2007 22:19:07 +0200 (Marcel Gruenauer <marcel@cpan.org>)
     - simplified accessor declaration in documentation

1.07  Thu, 18 Oct 2007 10:09:39 +0200 (Marcel Gruenauer <marcel@cpan.org>)
     - fixed version requirement of Test::More

1.06  Sun, 23 Sep 2007 10:03:20 +0200 (Marcel Gruenauer <marcel@cpan.org>)
     - now uses Module::Install::StandardTests

1.05  Sat, 08 Sep 2007 09:38:55 +0200 (Marcel Gruenauer <marcel@cpan.org>)
     - fixed the documentation
     - use warnings; use strict;
     - converted to new-style distro (with Module::Install and such)

1.04  Wed, 11 Jan 2006 15:44:53 +0100 (Marcel Gruenauer <marcel@cpan.org>)
     - Consistency improvements

1.03  Wed, 09 Feb 2005 15:26:03 +0100 (Marcel Gruenauer <marcel@cpan.org>)
     - Removed Test::Distribution requirement
     - Added method chaining

1.02  Fri, 06 Sep 2002 18:58:47 +0200 (Marcel Gruenauer <marcel@cpan.org>)
     - changed AUTOLOAD to return undef. Up to now it returned the glob created
       by *{$AUTOLOAD}, which is not what the called would expect from sub {}.

1.01  Sun, 18 Aug 2002 13:52:05 +0200 (Marcel Gruenauer <marcel@cpan.org>)
     - Class::Null is now a singleton to reduce memory overhead (thanks
       nwetters)
     - Class::Null now creates an empty subroutine when autoloading, to
       increase the speed of subsequent calls (they don't have to go through
       AUTOLOAD again) (thanks nwetters)

1.00  Sat, 17 Aug 2002 11:38:47 +0200 (Marcel Gruenauer <marcel@cpan.org>)
     - changed 'our' to 'use vars' for $VERSION so it runs under earlier
       versions of perl (thanks richardc)
     - fixed a POD typo
     - removed the '5.008' requirement from Makefile.PL

0.01  Fri, 16 Aug 2002 17:01:00 +0100 (Marcel Gruenauer <marcel@cpan.org>)
     - original version


lib/Class/Null.pm
use 5.008;
use strict;
use warnings;

package Class::Null;
BEGIN {
  $Class::Null::VERSION = '2.101420';
}
# ABSTRACT: Implements the Null Class design pattern
use overload
  'bool'   => sub { 0 },
  '""'     => sub { '' },
  '0+'     => sub { 0 },
  fallback => 1;
our $null = bless {}, __PACKAGE__;
sub AUTOLOAD { $null }
1;


__END__
=pod

=head1 NAME

Class::Null - Implements the Null Class design pattern

=head1 VERSION

version 2.101420

=head1 SYNOPSIS

  use Class::Null;

  # some class constructor and accessor declaration here

  sub init {
    my $self = shift;
    # ...
    $self->log(Class::Null->new);
    # ...
  }

  sub do_it {
    my $self = shift;
    $self->log->log(level => 'debug', message => 'starting to do it');
    # ...
    $self->log->log(level => 'debug', message => 'still doing it');
    # ...
    $self->log->log(level => 'debug', message => 'finished doing it');
  }

=head1 DESCRIPTION

This class implements the Null Class design pattern.

Suppose that methods in your object want to write log messages to a log
object. The log object is possibly stored in a slot in your object and
can be accessed using an accessor method:

  package MyObject;

  use base 'Class::Accessor';
  __PACKAGE__->mk_accessors(qw(log));

  sub do_it {
    my $self = shift;
    $self->log->log(level => 'debug', message => 'starting to do it');
    ...
    $self->log->log(level => 'debug', message => 'still doing it');
    ...
    $self->log->log(level => 'debug', message => 'finished doing it');
  }

The log object simply needs to have a C<log()> method that accepts
two named parameters. Any class defining such a method will do, and
C<Log::Dispatch> fulfills that requirement while providing a lot of
flexibility and reusability in handling the logged messages.

You might want to log messages to a file:

  use Log::Dispatch;

  my $dispatcher = Log::Dispatch->new;

  $dispatcher->add(Log::Dispatch::File->new(
    name      => 'file1',
    min_level => 'debug',
    filename  => 'logfile'));

  my $obj = MyObject->new(log => $dispatcher);
  $obj->do_it;

But what happens if we don't define a log object? Your object's methods
would have to check whether a log object is defined before calling the
C<log()> method. This leads to lots of unwieldy code like

  sub do_it {
    my $self = shift;
    if (defined (my $log = $self->log)) {
      $log->log(level => 'debug', message => 'starting to do it');
    }
    ...
    if (defined (my $log = $self->log)) {
      $log->log(level => 'debug', message => 'still doing it');
    }
    ...
    if (defined (my $log = $self->log)) {
      $log->log(level => 'debug', message => 'finished doing it');
    }
  }

The proliferation of if-statements really distracts from the actual call
to C<log()> and also distracts from the rest of the method code. There
is a better way. We can ensure that there is always a log object that we can
call C<log()> on, even if it doesn't do very much (or in fact, anything at
all).

This object with null functionality is what is called a null object. We can
create the object the usual way, using the C<new()> constructor, and call any
method on it, and all methods will do the same - nothing. (Actually, it
always returns the same C<Class::Null> singleton object, enabling method
chaining.) It's effectively a catch-all object. We can use this class with our
own object like this:

  package MyObject;

  use Class::Null;

  # some class constructor and accessor declaration here

  sub init {
    my $self = shift;
    ...
    $self->log(Class::Null->new);
    ...
  }

  sub do_it {
    my $self = shift;
    $self->log->log(level => 'debug', message => 'starting to do it');
    ...
    $self->log->log(level => 'debug', message => 'still doing it');
    ...
    $self->log->log(level => 'debug', message => 'finished doing it');
  }

This is only one example of using a null class, but it can be used whenever
you want to make an optional helper object into a mandatory helper object,
thereby avoiding unnecessarily complicated checks and preserving the
transparency of how your objects are related to each other and how they
call each other.

Although C<Class::Null> is exceedingly simple it has been made into a
distribution and put on CPAN to avoid further clutter and repetitive
definitions.

=head1 METHODS

=head2 new

Returns the singleton null object.

=head2 Any other method

Returns another singleton null object so method chaining works.

=head1 OVERLOADS

=over 4

=item Boolean context

In boolean context, a null object always evaluates to false.

=item Numeric context

When used as a number, a null object always evaluates to 0.

=item String context

When stringified, a null object always evaluates to the empty string.

=back

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Class-Null>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Class-Null/>.

The development version lives at
L<http://github.com/hanekomu/Class-Null/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2005 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



inc/Module/Install/StandardTests.pm
#line 1
package Module::Install::StandardTests;

use warnings;
use strict;
use File::Spec;

use base 'Module::Install::Base';

our $VERSION = '0.02';

sub use_standard_tests {
    my $self = shift;
    $self->build_requires('Test::More');
    $self->build_requires('UNIVERSAL::require');
    $self->write_standard_test_compile;
    $self->write_standard_test_pod;
    $self->write_standard_test_pod_coverage;
    $self->write_standard_test_perl_critic;
}


sub write_test_file {
    my ($self, $filename, $code) = @_;
    $filename = File::Spec->catfile('t', $filename);

    # Outdent the code somewhat. Remove first empty line, if any. Then
    # determine the indent of the first line. Throw that amount of indenting
    # away from any line. This allows you to indent the code so it's visually
    # clearer (see methods below) while creating output that's indented more
    # or less correctly. Smoke result HTML pages link to the .t files, so it
    # looks neater.

    $code =~ s/^ *\n//;
    (my $indent = $code) =~ s/^( *).*/$1/s;
    $code =~ s/^$indent//gm;

    print "Creating $filename\n";
    open(my $fh, ">$filename") or die "can't create $filename $!";

    my $perl = $^X;
    print $fh <<TEST;
#!$perl -w

use strict;
use warnings;

$code
TEST

    close $fh or die "can't close $filename $!\n";
    $self->realclean_files($filename);
}


sub write_standard_test_compile {
    my $self = shift;
    $self->write_test_file('000_standard__compile.t', q/
        use Test::More;
        eval "use Test::Compile";
        plan skip_all => "Test::Compile required for testing compilation" if $@;
        all_pm_files_ok();
    /);
}


sub write_standard_test_pod {
    my $self = shift;
    $self->write_test_file('000_standard__pod.t', q/
        use Test::More;
        eval "use Test::Pod";
        plan skip_all => "Test::Pod required for testing POD" if $@;
        all_pod_files_ok();
    /);
}


sub write_standard_test_pod_coverage {
    my $self = shift;
    $self->write_test_file('000_standard__pod_coverage.t', q/
        use Test::More;
        eval "use Test::Pod::Coverage";
        plan skip_all =>
            "Test::Pod::Coverage required for testing POD coverage" if $@;
        all_pod_coverage_ok();
    /);
}


sub write_standard_test_perl_critic {
    my $self = shift;
    $self->write_test_file('000_standard__perl_critic.t', q/
        use FindBin '$Bin';
        use File::Spec;
        use UNIVERSAL::require;
        use Test::More;

        plan skip_all =>
            'Author test. Set $ENV{TEST_AUTHOR} to a true value to run.'
            unless $ENV{TEST_AUTHOR};

        my %opt;
        my $rc_file = File::Spec->catfile($Bin, 'perlcriticrc');
        $opt{'-profile'} = $rc_file if -r $rc_file;

        if (Perl::Critic->require('1.078') &&
            Test::Perl::Critic->require &&
            Test::Perl::Critic->import(%opt)) {

            all_critic_ok("lib");
        } else {
            plan skip_all => $@;
        }
    /);
}


1;

__END__

#line 216



Makefile.PL

use strict;
use warnings;

 BEGIN { require 5.008; } 

use ExtUtils::MakeMaker 6.31;



my %WriteMakefileArgs = (
                       'test' => {
                                   'TESTS' => 't/*.t'
                                 },
                       'NAME' => 'Class::Null',
                       'DISTNAME' => 'Class-Null',
                       'CONFIGURE_REQUIRES' => {
                                                 'ExtUtils::MakeMaker' => '6.31'
                                               },
                       'AUTHOR' => 'Marcel Gruenauer <marcel@cpan.org>',
                       'BUILD_REQUIRES' => {
                                             'Test::More' => '0.94',
                                             'Scalar::Util' => '0',
                                             'parent' => '0',
                                             'English' => '0',
                                             'File::Find' => '0',
                                             'File::Temp' => '0',
                                             'Carp' => '0'
                                           },
                       'ABSTRACT' => 'Implements the Null Class design pattern',
                       'EXE_FILES' => [],
                       'VERSION' => '2.101420',
                       'PREREQ_PM' => {
                                        'overload' => '0'
                                      },
                       'LICENSE' => 'perl'
                     );


delete $WriteMakefileArgs{BUILD_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.56) };

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



